File: presentation/src/test/java/io/tripled/social/client/presentation/ReadEvalPrintLoopTest.java
package io.tripled.social.client.presentation;

import org.junit.Before;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class ReadEvalPrintLoopTest {

  private ReadEvalPrintLoop repl;
  private TestInput input;
  private TestOutput output;

  @Before
  public void setUpReplWithEchoCommand() {
    input = new TestInput();
    output = new TestOutput();
    repl = new ReadEvalPrintLoop(input, output, null, null, null, null);
  }

  @Test(expected = NullPointerException.class)
  public void throwsExceptionWithNullInput() {
    new ReadEvalPrintLoop(null, output, null, null, null, null);
  }

  @Test(expected = NullPointerException.class)
  public void throwsExceptionWithNullOutput() {
    new ReadEvalPrintLoop(new TestInput(), null, null, null, null, null);
  }

  @Test
  public void quits() {
    input.addInput("\\q");

    repl.run();

    assertThat(repl.isStopped(), is(true));
    output.assertIsEmpty();
  }

}

File: src/test/kotlin/io/tripled/social/client/infrastructure/cli/ReadEvalPrintLoopTest.kt
package io.tripled.social.client.infrastructure.cli

import io.tripled.social.client.application.usecase.spies.FollowUserSpy
import io.tripled.social.client.application.usecase.spies.PostMessageSpy
import io.tripled.social.client.application.usecase.spies.ReadMessagesSpy
import io.tripled.social.client.application.usecase.spies.ReadWallSpy
import io.tripled.social.client.infrastructure.cli.controller.FollowController
import io.tripled.social.client.infrastructure.cli.controller.PostController
import io.tripled.social.client.infrastructure.cli.controller.ReadController
import io.tripled.social.client.infrastructure.cli.controller.WallController
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test

internal class ReadEvalPrintLoopTest {
    private val input: TestInput = TestInput()
    private val output: TestOutput = TestOutput()
    private val repl: ReadEvalPrintLoop = ReadEvalPrintLoop(
        input, output,
        PostController(PostMessageSpy()),
        ReadController(ReadMessagesSpy()),
        WallController(ReadWallSpy()),
        FollowController(FollowUserSpy())
    )

    @Test
    fun quits() {
        input.addInput("\\q")

        repl.run()

        MatcherAssert.assertThat<Boolean?>(repl.isStopped, Matchers.`is`<Boolean?>(true))
        output.assertIsEmpty()
    }
}
